### ISSUE TITLE
Current implementation of patcher does not support sequence packing

### ISSUE THREAD

**Issue by isayoften** [facebookresearch/blt#36]
I am right, if i use sequence packing with custom masking (to not look cross document) i need to manually start new patch at BOS token to prevent situation where end of one document and start of next document fall into one patch?

For example: [bos, 1,2,3, eos, bos, 4,5,6,eos]. So, i don't want the situation like [1,0,0,0,0,_0_,1,0,0,0] to happen. 

Also, what is the purpose of _include_next_token_ arg in patcher?

**Comment by Vectorrent** [comment]
This is a good question, and one I hadn't considered. I use sequence packing as well.

The way I see it, we can probably set the position of our special tokens to extremely large entropy values, and thereby force the patcher to split on those boundaries:
```py
def patch_entropies_for_special_tokens(
    input_ids: torch.LongTensor,
    entropy_scores: torch.Tensor,
    special_token: int = 0,
    high_entropy_value: float = 1e9,
) -> torch.Tensor:
    """
    Modifies entropy scores to force patch boundaries at special tokens.

    Args:
        input_ids: Input tensor of shape [batch_size, seq_len]
        entropy_scores: Entropy scores of shape [batch_size, seq_len]
        special_token: The special_token value (default: 0)
        high_entropy_value: Value to set for special token positions (default: 1e9)
    """
    # Create special token mask: True where special token exists
    token_mask = input_ids == special_token

    # Where token_mask is True, use high_entropy_value
    # Where token_mask is False, keep original entropy_scores
    modified_entropy_scores = torch.where(
        token_mask,
        torch.tensor(
            high_entropy_value,
            device=entropy_scores.device,
            dtype=entropy_scores.dtype,
        ),
        entropy_scores,
    )

    return modified_entropy_scores
```
From there, we'll have to create block-causal masks and handle the positional encoding differently, in the encoder, and in the entropy model. I don't think any of the current BLT code supports this, so it would all be custom work.

The one part I haven't figured out yet, is where to set the sequence boundaries after the downsampling operation? All the way up to that point, it's fairly easy to discern - because the sequence length has not changed. But after the downsampling... I'm not sure where to split the block-causal masking at. I suppose it's going to require splitting the patch_ids into different groups, depending upon where the special tokens are located. I haven't figured this part out yet, but I'll report back if I do.

**Comment by isayoften** [comment]
We need entropy to be high one step before BOS, because only after getting high entropy patcher will assign "1" for the next byte meaning new patch will start at the next byte. I made it this way (there are also some changes for my purposes, you may look only at bos_mask part):

```
def find_entropy_patch_start_ids(entropies, tokens, threshold, bos_id):
    bs, seq_len = entropies.shape[:2]

    first_ids = torch.tensor([0, 1], dtype=torch.long, device=entropies.device).unsqueeze(0).repeat(bs, 1)
    preds_truncation_len = first_ids.shape[1]  # remove the first preds because they will be start of patches.
    entropies = entropies[:, 1:]

    patch_start_mask = entropies > threshold

    # we need to shift mask to the left because patch starts "look" at one position forward
    bos_mask = torch.cat(
        ((tokens[:, 1:] == bos_id)[:, 1:], torch.tensor([[False]], dtype=torch.bool, device=tokens.device)), dim=1
    )
    patch_start_mask = patch_start_mask | bos_mask

    patch_start_mask = patch_start_mask[:, :-1]

    patch_start_ids = patch_start_ids_from_patch_start_mask(patch_start_mask)

    patch_start_ids = torch.cat((first_ids, patch_start_ids + preds_truncation_len), dim=1)
    return patch_start_ids
```

**Comment by Vectorrent** [comment]
![Image](https://github.com/user-attachments/assets/a7e48779-df4a-4fe4-8391-f388a5d05c18)

This seems to suggest that the local block causal attention masks do, in fact, prevent cross-document attention. I cannot verify it actually works though, since I don't use this part of the code.

**Comment by isayoften** [comment]
I think the appropriate mask for patches is created here:

<img width="692" alt="Image" src="https://github.com/user-attachments/assets/0a023b8c-4275-4104-9c11-c2e5f7761789" />
